api/service/legacysync/epoch_syncing.go

```diff
@@ -0,0 +1,240 @@
+package legacysync
+
+import (
+	"sync"
+	"time"
+
+	"github.com/Workiva/go-datastructures/queue"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/harmony-one/harmony/consensus"
+	"github.com/harmony-one/harmony/core"
+	"github.com/harmony-one/harmony/core/types"
+	"github.com/harmony-one/harmony/internal/chain"
+	"github.com/harmony-one/harmony/internal/utils"
+	"github.com/harmony-one/harmony/node/worker"
+	"github.com/harmony-one/harmony/p2p"
+	"github.com/harmony-one/harmony/shard"
+	"github.com/pkg/errors"
+)
+
+type EpochSync struct {
+	beaconChain        *core.BlockChain
+	selfip             string
+	selfport           string
+	selfPeerHash       [20]byte // hash of ip and address combination
+	commonBlocks       map[int]*types.Block
+	lastMileBlocks     []*types.Block // last mile blocks to catch up with the consensus
+	syncConfig         *SyncConfig
+	isExplorer         bool
+	stateSyncTaskQueue *queue.Queue
+	syncMux            sync.Mutex
+	lastMileMux        sync.Mutex
+
+	syncStatus syncStatus
+}
+
+// GetSyncStatus get the last sync status for other modules (E.g. RPC, explorer).
+// If the last sync result is not expired, return the sync result immediately.
+// If the last result is expired, ask the remote DNS nodes for latest height and return the result.
+func (ss *EpochSync) GetSyncStatus() SyncCheckResult {
+	return ss.syncStatus.Get(func() SyncCheckResult {
+		return ss.isInSync(false)
+	})
+}
+
+// isInSync query the remote DNS node for the latest height to check what is the current
+// sync status
+func (ss *EpochSync) isInSync(doubleCheck bool) SyncCheckResult {
+	if ss.syncConfig == nil {
+		return SyncCheckResult{} // If syncConfig is not instantiated, return not in sync
+	}
+	otherHeight1 := getMaxPeerHeight(ss.syncConfig)
+	curBlock := ss.beaconChain.CurrentBlock()
+	epochLastBlock := shard.Schedule.EpochLastBlock(curBlock.Epoch().Uint64())
+	lastHeight := ss.beaconChain.CurrentBlock().NumberU64()
+
+	if !doubleCheck {
```

Comment:
Whether the epoch chain is in sync should be determined by this:

1. Get my latest epoch block (note since it's epoch chain, the CurrentBlock() will return an epoch block (the last block of an epoch). The epoch of this block is my_epoch.
2. Get the highest block from Peer. The epoch of this block is their_epoch.
3. Compare whether my_epoch == their_epoch-1.  Note it's almost impossible to have my_epoch == their_epoch because I only have the epoch block which is the last block of an epoch. So when I sync'ed the epoch block of an epoch, the beacon chain should already moved to the next epoch.

Reply:
just a side node on this comment @rlan35, it happened on the mainnet i am testing that we have this logs constantly :
```
{"level":"info","caller":"/home/sp/harmony/harmony/api/service/legacysync/epoch_syncing.go:121","time":"2022-06-19T18:56:04.933824999-05:00","message":"[EPOCHSYNC] Node is OUT OF SYNC (isBeacon: true, ShardID: 0, otherHeight: 18446744073709551615, currentHeight: 27951103)"}
```
until a restart of the node is initiated, and the otherHeight is correct

Reply:
18446744073709551615 is max value for uint64

It started with this value, other peers should overwrite it, but it doesn't. Seems like no other peers or no even one peer responses correctly.